File: simple-client.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>
#include <cstring>

// Global variable for registration status
bool isRegistered = false;

// Registration handler - called when server registers the client
void handleRegistration(bool isAccepted) {
    if (isAccepted) {
        isRegistered = isAccepted;
        std::cout << "✅ Client registered successfully: " << isRegistered << std::endl;
    }
}

// Server notification handler - logs when called
void handleServerNotification(const char* data) {
    std::cout << "📨 Server notification received: " << data << std::endl;
}

// Function to send periodic notifications
void sendPeriodicNotifications(void* clientPtr) {
    while (true) {
        if (isRegistered) {
            std::string message = R"({
                "type": "heartbeat",
                "message": "Hello from simple client",
                "timestamp": ")"
                + std::to_string(std::time(nullptr)) + "\"}";
            
            sendNotification(clientPtr, message.c_str());
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
}

// Function to send periodic logs
void sendPeriodicLogs(void* clientPtr) {
    while (true) {
        if (isRegistered) {
            std::string logData = R"({
                "level": "info",
                "message": "Periodic log message",
                "timestamp": ")"
                + std::to_string(std::time(nullptr)) + "\"}";
            
            sendLog(clientPtr, logData.c_str());
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
}

int main() {
    std::cout << "🚀 Simple MQTT client started" << std::endl;
    std::cout << "📤 Will send notifications and logs every 5 seconds after registration" << std::endl;
    
    // Create MQTT client
    void* client = createMQTTClient("simple_client_001", handleRegistration, handleServerNotification);
    
    if (!client) {
        std::cerr << "❌ Failed to create MQTT client" << std::endl;
        return 1;
    }
    
    // Start periodic sending in background threads
    std::thread notificationThread(sendPeriodicNotifications, client);
    std::thread logThread(sendPeriodicLogs, client);
    
    // Keep main thread alive
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    
    // Cleanup (this won't be reached in this simple example)
    cleanupMQTTClient(client);
    return 0;
}

File: simple-client.js
import { createMQTTClient } from './node-client.js';

// Global variable for registration status
let isRegistered = false;

// Registration handler - called when server registers the client
const handleRegistration = (isAccepted) => {
    if (isAccepted === true) {
        isRegistered = isAccepted;
        console.log('✅ Client registered successfully', isRegistered);
    }
};

// Server notification handler - logs when called
const handleServerNotification = (data) => {
    console.log('📨 Server notification received:', JSON.stringify(data));
};

// Create MQTT client
const client = createMQTTClient('simple_client_003', handleRegistration, handleServerNotification);

// Send notification every 10 seconds
setInterval(() => {
    if (isRegistered) {
        client.sendNotification({
            "data": {
              "AuthenticationModule": "kerberos",
              "DestinationDeviceAddress": "172.16.100.20",
              "DestinationDeviceName": "SR2019",
              "DomainControllerAddress": "172.16.100.10",
              "DomainControllerName": "DC-2",
              "EncryptionType": "0x12 (AES-256-CTS-HMAC-SHA1-96)",
              "NotifiedAgentName": "dc-2.fuck.local",
              "NotifiedAgentType": "active directory,radius",
              "RequestedService": "TERMSRV",
              "SourceDeviceAddress": "172.16.100.1",
              "SourceDeviceName": "MDFEETR",
              "domain": "fuck",
              "notifId": Math.floor(10000 + Math.random() * 90000).toString(),
              "targetedUser": "s",
              "targetedUserSid": "S-1-5-21-141073069-1336069126-459662567-501",
              "targetedUserSidX": 4,
              "timestamp": "2025-10-01T18:05:25+03:30",
              "LogonLevel": "Interactive"
            },
            "type": "NOTIFICATION_REQUEST"
          });
        console.log('🔥 Notification sent');
    }
}, 250000);

// Send notification every 10 seconds
// setInterval(() => {
//     if (isRegistered) {
//         client.sendNotification({
//             "data": {
//               "AuthenticationModule": "kerberos",
//               "DestinationDeviceAddress": "172.16.100.20",
//               "DestinationDeviceName": "DC-2$",
//               "DomainControllerAddress": "172.16.100.10",
//               "DomainControllerName": "DC-2",
//               "EncryptionType": "0x12 (AES-256-CTS-HMAC-SHA1-96)",
//               "NotifiedAgentName": "dc-2.fuck.local",
//               "NotifiedAgentType": "active directory,radius",
//               "RequestedService": "TERMSRV",
//               "SourceDeviceAddress": "172.16.100.1",
//               "SourceDeviceName": "MDFEETR1",
//               "domain": "fuck",
//               "status": "allowed",
//               "notifId": Math.floor(10000 + Math.random() * 90000).toString(),
//               "targetedUser": "s1",
//               "targetedUserSid": "S-1-5-21-141073069-1336069126-459662567-501",
//               "targetedUserSidX": 4,
//               "timestamp": "2025-10-01T18:05:25+03:30",
//               "LogonLevel": "Interactive"
//             },
//             "type": "NOTIFICATION_POLICY"
//           });
//         console.log('🔥 Notification Policy sent');
//     }
// }, 5000);


// // Send log every 10 seconds
// setInterval(() => {
//     if (isRegistered) {
//         client.sendLog({
//             level: 'info',
//             message: 'Periodic log message',
//             test: 1
//         });
//         console.log('🔥 Log sent');
//     }
// }, 25000);

console.log('🚀 Simple MQTT client started');
console.log('📤 Will send notifications and logs every 10 seconds after registration');
